### ISSUE TITLE
High frequency messages

### ISSUE THREAD

**Issue by mgazic** [MilesEngineering/ucplatform#1]
There's some common use cases where an application might need high speed data, that we want to avoid sending to every other consumer at high speed.  That can be accomplished in an ad hoc manner in application code right now, but it'd be better to have support in ucplatform.  This could be done in a few different ways:

1) Implement decimation in subscriptions.  Either one-out-of-N or time based (every N milliseconds, N=zero or -1 for "i want everything").  It'd be a feature of the subscription that a publisher has to look at when deciding who to queue data to.  There's no obvious way for it to work for 'greedy' consumers that wants everything, because the most straight-forward approach is for each subscription object to have metadata (counter, or timestamp of last consumption) for each message type.  Perhaps we could have greedy subscribers *also* add decimated subscriptions, and if they get it specifically, they won't get it greedily.  That would let a subscriber get everything at full rate except the few specific messages they decide to decimate.  If we do decide to use the general message passing mechanism for high speed data, we might need to optimize it to work for 100+Hz data, at least from one producer (often an ISR) to one consumer (often the main G&C task).
2) Make an alternate high-speed point-to-point data path that can transfer messages.  This would work well for a G&C application that needs full-rate data for processing of sensor data or commands, but would itself decimate that data before publishing it.

When investigating optimizing the pubsub mechanism, consider using other FreeRTOS APIs ([regular, alternative, and lightweight](https://www.freertos.org/message_passing_performance.html)).  Probably we don't need super low interrupt latency, so doing critical sections around the lightweight ("fromISR") API would work best.  Also consider replacing the MessageQueue used a "free list" in the MessagePool with a circular buffer protected by critical sections, because nobody ever needs to block when they store or fetch free buffers, because the pool is big enough to hold all the buffers that exist, and a critical section around a circular buffer would be very fast.  Currently every message needs to be allocated, sent/received, and freed, and optimizing the allocation and free would probably be easier than optimizing the send/receive (which do need to wake a thread, hence have to use FreeRTOS APIs).  Send/receive could perhaps be implemented as a critial-section protected circular buffer with [Task Notifications](https://www.freertos.org/RTOS-task-notifications.html), which are considerably faster than other FreeRTOS primitives, and their only restrictions are they can only have a single recipient (which is fine for MessageQueue/MessageClient), and cannot have a sender block (which is fine as well, because we already try to size the existing queue to be as large as the entire pool, and we don't let senders block).

When considering a high-speed point-to-point data path, consider avoiding FreeRTOS entirely, and having a circular buffer wrapped in critical sections that the reader will poll from at a fixed rate.  That'd use a lot of message buffers though, so also consider a copy-based mechanism that only stores the latest data, and has set/get functions that copy in and out in critical sections.  In any case, it should be Message oriented, and should avoid churning on allocate and free.  For the copy-based version that stores the latest data, this could be done with a templated MessageStore class that doesn't require declaring/destroying Message objects, and instead just copies from one external message (that is persistent in the writer) to the internal copy, and from the internal copy to another external message (that is persistent in the reader).

Further notes:
```
 * The source code that implements the alternative (Alt) API is much
 * simpler because it executes everything from within a critical section.
 * This is the approach taken by many other RTOSes, but FreeRTOS.org has the
 * preferred fully featured API too.  The fully featured API has more
 * complex code that takes longer to execute, but makes much less use of
 * critical sections.  Therefore the alternative API sacrifices interrupt
 * responsiveness to gain execution speed, whereas the fully featured API
 * sacrifices execution speed to ensure better interrupt responsiveness.
```

Using the "lightweight" API for the freelist could be as simple as:
Allocate:
```
    taskENTER_CRITICAL();
        xQueueReceiveFromISR( xQueue, pvBuffer, &x );
    taskEXIT_CRITICAL();
    // nothing needs to be done with xYieldRequired because nobody can pend on writing to the free list
```
Free:
```
    taskENTER_CRITICAL();
        xYieldRequired = xQueueSendToFrontFromISR( xQueue, pvItemToQueue, pdFALSE );
    taskEXIT_CRITICAL();
    // nothing needs to be done with xYieldRequired because nobody can pend on reading the free list
```
Look at details [here](https://www.freertos.org/message_passing_performance.html#light), specifically `Light weight example 3 - introducing preemption and interrupt safety`.

**Comment by mgazic** [comment]
Here's a rough hack at code that optimizes sending messages in various ways:
```
Look at issue related to high-rate message traffic:
https://github.com/MilesEngineering/ucplatform/issues/1

We need a solution ASAP that works for VectorNav (AP rate*2?) and SBUS data (50Hz?)

Note:

nobody should ever block on the free list
    if it's empty, a reader is out of luck, and need to continue on
    it can never be full when writing to it, because there's as many slots as items from the Pool


currently there's multiple readers and multiple writers of a MessagePool (aka free list), and nobody can block when reading or writing it
currently there's multiple writers to a MessageClient, but only one reader (the client itself, which is a Task)
    some readers want to block and be woken when a write occurs
    other readers might want to poll their queue, and they'll sleep for a fixed time between polling

since no blocking should occur with MessagePool, it doesn't need to use a freertos queue, it can use an
access-protected circular buffer that just briefly enters a critical section

for readers that don't want to block on their queue and instead will poll it, their queue could be
implemented with a circular buffer and critical sections
    if they don't poll fast enough, they could exhaust the Pool of available buffers by holding on to all of them.
        so maybe they need a separate Pool, just for their messages.

for readers that want to block on their queue
    we could keep using a freertos queue. could do xQueueReceive(), or xQueueReceiveFromISR() inside critical section

---- Below is some example code that has a more optimized version of message transfer for:
1) Storage of only the single latest copy of a message, and data is copied in and out
2) A single-producer single-consumer queue that does not perform data copies, and is suitable for ISR->Task or Task->ISR
3) A more general multi-producer queue, that can handle either multiple non-blocking consumers (like MessagePool), or exactly one blocking consumer.
#define taskENTER_CRITICAL_FROM_ANYWHERE \
    do{\
        if(xPortIsInsideInterrupt()) { taskENTER_CRITICAL_FROM_ISR(); }\
        else{ taskENTER_CRITICAL();} \
    } while(0)

#define taskEXIT_CRITICAL_FROM_ANYWHERE \
    do{\
        if(xPortIsInsideInterrupt()) { taskEXIT_CRITICAL_FROM_ISR(); }\
        else{ taskEXIT_CRITICAL();} \
    } while(0)

// Used to store a single copy of the latest value of a message.
// Can be used between an ISR and a Task, when the ISR is reading
// data, and the Task only wants the latest data it hasn't seen yet.
class LatestMessageStorage
//# Or MessageStorage, SingleMessageStorage, NewestMessageStorage?
//# Or maybe pick a name that makes it clear that this is intended
//# for ISR->Thread communication?
{
    public:
        LatestMessageStorage(MessageBuffer* buffer)
        : m_mb(buffer)
        {
        }
        // Functions are purposefully named read/write to try to emphasize that
        // this sets/gets one value, instead of acting like a stream/pipe/queue
        void write(MessageBuffer& mb)
        {
            taskENTER_CRITICAL_FROM_ANYWHERE();
            mb->copy(*m_mb);
            m_data_available = true;
            taskEXIT_CRITICAL_FROM_ANYWHERE();
        }
        bool read(MessageBuffer& mb)
        {
            if(!m_data_available)
            {
                return false;
            }
            taskENTER_CRITICAL_FROM_ANYWHERE();
            m_mb->copy(mb);
            m_data_available = false;
            taskEXIT_CRITICAL_FROM_ANYWHERE();
            return true;
        }
    private:
        MessageBuffer* m_mb;
        bool m_data_available = false;
};

// Single-producer single-consumer queue, implemented without blocking.
// Useful for sending a stream of messages between a single ISR and a single
// thread.  If the ISR operates at a very high data rate, you probably should
// consider using a separate MessagePool for it, so that if it runs out of
// message buffers, the entire system isn't starved of message buffers.
template <uint8_t SIZE>
class SingleProducerSingleConsumerMessageQueue
//# Or DirectMessageQueue, PolledMessageQueue?
//# I wouldn't want to call it NonBlockingMessageQueue because that has connotations of
//# not blocking the sender, which none of my implementations of a MessageQueue would ever
//# do.  This one is non-blocking for the *receiver*, which is actually a bad thing because
//# it doesn't include the feature of them being able to do a blocking read that will wake
//# them up instantly when data is availble, so they have to Poll for data instead.
{
    public:
        bool put(MessageBuffer* mb)
        {
            uint8_t head = m_head + 1;
            if(head == SIZE)
            {
                head = 0;
            }
            if(head == m_tail)
            {
                // Might be better to drop old messages instead of new ones
                // when the queue is full, but that would require use of
                // critical sections, because the writer would modify the head
                // index instead of just the tail index.
                m_overflow_count++;
                return False;
            }
            buf[++last] = mb;
            return True;
        }
        MessageBuffer* get()
        {
            if(m_head == m_tail)
            {
                return nullptr;
            }
            MessageBuffer* ret = buf[m_tail];
            uint8_t tail = m_tail + 1;
            if(tail == SIZE)
            {
                tail = 0;
            }
            m_tail = tail;
            return ret;
        }
        MessageBuffer* buf[SIZE];
        m_overflow_count=0;
        uint8_t m_head=0;
        uint8_t m_tail=0;
};

// Multiple-producer queue, with either just a single blocking consumer, or
// multiple consumers if they are all non-blocking.
// Uses FreeRTOS Notifications (https://www.freertos.org/RTOS-task-notification-API.html)
// to add optional blocking behaviour around a single-producer single-consumer queue.
// Also uses critical sections to be thread-safe which works for both a thread doing
// the reading (blocking or non-blocking), or a ISR.
const void* MessageQueue::s_wakeSentinel = (void*)~0;
template <uint8_t SIZE>
class MultiProducerSingleBlockingConsumerMessageQueue
//# Or if this replaces the default implementation, just "MessageQueue"?
{
    public:
        MultiProducerSingleBlockingConsumerMessageQueue(TaskHandle_t owner)
        : m_owner(owner)
        {
        }
        bool put(MessageBuffer* msgbuf)
        {
            // Increment the refcount for the recipient before we send it to them,
            // so they can decrement it as soon as they're done with it, with no
            // worry about them decrementing before we increment if they're higher
            // priority than the sender and would preempt the sender.
            if(msgbuf != &s_wakeSentinel)
            {
                msgbuf->increment_refcount();
            }
            bool ret;
            // since multiple threads and ISRs can write to the same queue, we
            // need to:
            // 1) enter a critical section
            // 2) write to the queue
            // 3) exit critical section
            // 4) notify the task that was waiting.
            // 5) if in an ISR, cause the scheduler to run if someone was woken.
            // the critical section.
            if(xPortIsInsideInterrupt())
            {
                BaseType_t xHigherPriorityTaskWoken = pdFALSE;
                taskENTER_CRITICAL_FROM_ISR();
                ret = queue.put(msgbuf);
                taskEXIT_CRITICAL_FROM_ISR();
                // if there's an owner, and the put() succeeded, wake the owner up.
                if(m_owner && ret)
                {
                    xTaskNotifyGiveFromISR(m_owner, &xHigherPriorityTaskWoken);
                }
                //# On ARM Cortex, you can call taskYIELD_FROM_ISR from anywhere in
                //# an ISR, and it'll cause a scheduler interrupt to occur after the
                //# ISR exits.  On other platforms, it'll directly cause a task switch,
                //# so it should only be done at the very end of an ISR.
                if(xHigherPriorityTaskWoken)
                {
                    taskYIELD_FROM_ISR();
                }
            }
            else
            {
                taskENTER_CRITICAL();
                ret = queue.put(msgbuf);
                taskEXIT_CRITICAL();
                if(ret)
                {
                    xTaskNotifyGive(m_owner);
                }
            }
            if(!ret && msgbuf != &s_wakeSentinel)
            {
                // If sending failed, decrement the reference count because the recipient won't
                // get the message and can't decrement the refcount when they're done with it.
                msgbuf->decrement_refcount();
            }
            return ret;
        }
        // This has a timeout so users in a Task can block on getting data.
        MessageBuffer* get(TickType_t wait_time=0)
        {
            MessageBuffer* ret = nullptr;
            if(xPortIsInsideInterrupt())
            {
                configASSERT(wait_time==0);
                taskENTER_CRITICAL_FROM_ISR()
                MessageBuffer* ret = queue.get();
                taskEXIT_CRITICAL_FROM_ISR();
            }
            else
            {
                // Only Queues that have an owner can be blocked on, because if
                // there's no owner set, then no writers can notify an owner.
                if(m_owner && wait_time)
                {
                    ulTaskNotifyTake(pdFalse, wait_time);
                }
                taskENTER_CRITICAL();
                MessageBuffer* ret = queue.get();
                taskEXIT_CRITICAL();
            }
            return ret;
        }
        // call wake() to wake up someone pending on a message queue, without
        // sending them valid data.
        void wake()
        {
            put(&s_wakeSentinel);
        }
        // The wake sentinel is a variable that we send the address of to wake
        // someone waiting on a message.  
        static const void* s_wakeSentinel;
    private:
        SingleProducerSingleConsumerMessageQueue<SIZE> queue;
        TaskHandle_t m_owner;
};

```

**Comment by mgazic** [comment]
Added MessageDecimator, which takes a Queue and a timeout.  You can give it a stream of messages and every timeout ms it will Publish, and in between that it will send the messages to a specific Queue.

This works for now, but in the future it'd be nice to have subscriptions have a timeout, so that the sender of a message doesn't have to use Decimator to avoid publishing, and doesn't have to know who the one queue is that wants full rate data.  That will also allow subscriptions to change via configuration messages, so that PC tools could decide what rate they want messages at (if not the default rate which would generally be lower to avoid saturating the USB, CAN, or TT&C links).

**Comment by mgazic** [comment]
Added a compile-time option to have the MessagePool's free list be a protected CircularBuffer, instead of a FreeRTOS queue.

Still need to explore changing to FreeRTOS notifications using a circular buffer instead of FreeRTOS queues in MessageQueue for clients to pend on receiving messages.
